core/lib/zksync_core/src/consensus/storage/mod.rs

```diff
@@ -259,14 +280,47 @@ impl PersistentBlockStore for BlockStore {
             .context("genesis is missing")?)
     }
 
-    async fn last(&self, ctx: &ctx::Ctx) -> ctx::Result<Option<validator::CommitQC>> {
-        self.inner
-            .access(ctx)
+    async fn state(&self, ctx: &ctx::Ctx) -> ctx::Result<BlockStoreState> {
+        let mut conn = self.inner.access(ctx).await.wrap("access()")?;
+
+        // Fetch the range of miniblocks in storage.
+        let block_range = conn.block_range(ctx).await.context("block_range")?;
+
+        // Fetch the range of certificates in storage.
+        let genesis = conn
+            .genesis(ctx)
             .await
-            .wrap("access()")?
+            .wrap("genesis()")?
+            .context("genesis missing")?;
+        let first_cert = genesis.fork.first_block.max(block_range.start);
+        let last_cert = conn
             .last_certificate(ctx)
             .await
-            .wrap("last_certificate()")
+            .wrap("last_certificate()")?;
+        let next_cert = last_cert
+            .as_ref()
+            .map_or(first_cert, |cert| cert.header().number.next());
+
+        // Check that the first certificate in storage has the expected miniblock number.
+        if let Some(got) = conn
+            .first_certificate(ctx)
+            .await
+            .wrap("first_certificate()")?
+        {
+            if got.header().number != first_cert {
```

Comment:
Dumb question: Is there a possibility to have certificates older than `first_cert`? If so, is there any harm from such certificates?

Reply:
Currently there is none, and this check ensures that we don't introduce such a bug.